src/openrct2/ReplayManager.cpp

```diff
@@ -0,0 +1,775 @@
+        GameAction::Ptr action;
+        uint8_t playerid = 0;
+        uint8_t callback = 0;
+        uint32_t commandIndex = 0;
+
+        bool operator<(const ReplayCommand& comp) const
+        {
+            // First sort by tick
+            if (tick < comp.tick)
+                return true;
+            if (tick > comp.tick)
+                return false;
+
+            // If the ticks are equal sort by commandIndex
+            return commandIndex < comp.commandIndex;
+        }
+    };
+
+    struct ReplayRecordData
+    {
+        MemoryStream parkData;
+        MemoryStream spriteSpatialData;
+        MemoryStream parkParams;
+        std::string name;      // Name of play
+        uint64_t timeRecorded; // Posix Time.
+        uint32_t tickStart;    // First tick of replay.
+        uint32_t tickEnd;      // Last tick of replay.
+        std::multiset<ReplayCommand> commands;
+        std::vector<std::pair<uint32_t, rct_sprite_checksum>> checksums;
+        uint32_t checksumIndex;
+    };
+
+    class ReplayManager final : public IReplayManager
+    {
+        enum class ReplayMode
+        {
+            NONE = 0,
+            RECORDING,
+            PLAYING,
+            NORMALISATION,
+        };
+
+    public:
+        virtual ~ReplayManager()
+        {
+        }
+
+        virtual bool IsReplaying() const override
+        {
+            return _mode == ReplayMode::PLAYING;
+        }
+
+        virtual bool IsRecording() const override
+        {
+            return _mode == ReplayMode::RECORDING;
+        }
+
+        virtual bool IsNormalising() const override
+        {
+            return _mode == ReplayMode::NORMALISATION;
+        }
+
+        virtual void AddGameCommand(
+            uint32_t tick, uint32_t eax, uint32_t ebx, uint32_t ecx, uint32_t edx, uint32_t esi, uint32_t edi, uint32_t ebp,
+            uint8_t callback) override
+        {
+            if (_currentRecording == nullptr)
+                return;
+
+            uint32_t args[7];
+            args[0] = eax;
+            args[1] = ebx;
+            args[2] = ecx;
+            args[3] = edx;
+            args[4] = esi;
+            args[5] = edi;
+            args[6] = ebp;
+
+            _currentRecording->commands.emplace(gCurrentTicks, args, callback, _commandId++);
+
+            // Force a checksum record the next tick.
+            _nextChecksumTick = tick + 1;
+        }
+
+        virtual void AddGameAction(uint32_t tick, const GameAction* action) override
+        {
+            if (_currentRecording == nullptr)
+                return;
+
+            MemoryStream stream;
+            DataSerialiser dsOut(true, stream);
+            action->Serialise(dsOut);
+
+            std::unique_ptr<GameAction> ga = GameActions::Create(action->GetType());
+            ga->SetCallback(action->GetCallback());
+
+            stream.SetPosition(0);
+            DataSerialiser dsIn(false, stream);
```

Comment:
the `true`/`false` have no meaning when reading this review. It should rather be `E_SAVING`/`E_READING` or something along those lines, see my comment below.